{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Can we use bioinformatics to tell how closely related pathogens are?\n",
    "\n",
    "In bioinformatics, we (the scientists) often want to compare many different bacteria, viruses, and fungi to understand their related functions. Think about bioinformatics as analyzing the blueprints for all organisms, allowing us to understand more throughly how things work down to the most basic building blocks of life.\n",
    "\n",
    "Bioinformaticians use many tools to understand the organisms they work with. Today, we will use k-mer distance estimation, a powerful analysis technique that can be used to estimate how similar multiple organism are to one another. By breaking down the genetic sequences of these organisms into smaller fragments known as k-mers, we can find patterns across the genome that make organisms act uniquely, affecting how an organism can act as a pathogen, as well as how it adapts to its environment.\n",
    "\n",
    "In this module, we will compare several different bacteria and viruses to get an understanding of how closely related these pathogens are. Background on some common human pathogens is found below:\n",
    "\n",
    "### E. coli (Escherichia coli):\n",
    "E. coli is a bacteria commonly found in the human GI tract. Some types of E. coli have a gene for a toxin that can make us sick, especially if found in contaminated food or water.\n",
    "\n",
    "### S. aureus (Staphylococcus aureus):\n",
    "S. aureus is a bacteria commonly found on human skin. Some types of S. aureus, like MRSA, have become resistant to many antibiotics, and can cause serious skin infections.\n",
    "\n",
    "### S. pyogenes (Streptococcus pyogenes):\n",
    "S. pyogenes, also known as Group A strep, is the main cause of strep throat, and can typically be treated with penicllin.\n",
    "\n",
    "### SARS-CoV-2 (Severe Acute Respiratory Syndrome Coronavirus 2):\n",
    "SARS-CoV-2 is the virus that causes COVID-19, and is typically spread through respiratory droplets.\n",
    "\n",
    "### H3N2 (Influenza H3N2 virus):\n",
    "H3N2 is the virus that causes influenza type A in humans, but can also affect birds and other mammals. H3N2 is also spread primarily through respiratory droplets. The H3N2 mutates frequently, warranting development of yearly flu shots aimed at the predicted flu subtype.\n",
    "\n",
    "# ***********************************************************"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Downloading the NCBI RefSeq Databse\n",
    "\n",
    "NCBI is a centralized hub for genomic data. The NCBI RefSeq database contains genomes of tens of thousands of organisms, carefully curate for public use.\n",
    "\n",
    "Our first step is to download the RefSeq database.\n",
    "\n",
    "Click on the cell below and hit the play button to run the cell of code. If the code runs successfully, we should see output saying so."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import urllib.request\n",
    "import gzip\n",
    "\n",
    "os.makedirs(\"genomes\", exist_ok=\"True\")\n",
    "\n",
    "# Downloads the latest refseq assembly summary if not already downloaded\n",
    "#if not os.path.exists(\"assembly_summary_refseq.txt\"):\n",
    "    #urllib.request.urlretrieve(\"https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_refseq.txt\", \"assembly_summary_refseq.txt\")\n",
    "    \n",
    "print(\"NCBI RefSeq database successfully downloaded\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Retrieving our Pathogen Sequences from RefSeq\n",
    "\n",
    "We have successfully downloaded the RefSeq database. However, we need to query our database to retrieve our genomes of interest. When the text box pops up, type in Escherichia coli, Staphylococcus aureus, Streptococcus pyogenes, Severe Acute Respiratory Syndrome Coronavirus 2, and Influenza H3N2 virus one by one. Copy and paste if needed since the database will not be able to find a genome in the case of a spelling mistake.\n",
    "\n",
    "Click on the cell below and hit the play button to run the cell of code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def query_database():\n",
    "    # Retrieves link from RefSeq database\n",
    "    with gzip.open(\"assembly_summary_refseq.txt.gz\", \"rt\") as species_database:\n",
    "        num_genomes = 0\n",
    "        while True:\n",
    "            # This handles user input, and stops looping if the user enters nothing\n",
    "            user_input = input(\"\"\"Enter a genome you'd like to download and hit enter:\n",
    "Once you have all the genomes you want, hit enter to continue.\\n\"\"\")\n",
    "            if not user_input:\n",
    "                break\n",
    "            \n",
    "            genus, species = user_input.lower().split(\" \")[:2]\n",
    "\n",
    "            # Parse the RefSeq database to find our organism of interst\n",
    "            link = \"\"\n",
    "            for line in species_database:\n",
    "                if line.startswith(\"#\"):\n",
    "                    continue\n",
    "                    \n",
    "                organism = line.split(\"\\t\")[0].lower()\n",
    "                if genus in organism and species in organism:\n",
    "                    link = line.split(\"\\t\")[1]\n",
    "                    break\n",
    "                    \n",
    "            if not link:\n",
    "                print(f\"No matches found for {genus} {species}\")\n",
    "                continue\n",
    "            \n",
    "            num_genomes += 1\n",
    "            \n",
    "            # Gets file name\n",
    "            file_name = link.split(\"/\")[-1] + \"_genomic.fna.gz\"\n",
    "\n",
    "            # Gets genome file download link\n",
    "            fna = f\"{link}/{file_name}\"\n",
    "\n",
    "            # Downloads, moves, and unzips file\n",
    "            urllib.request.urlretrieve(fna, os.path.join(\"genomes\", file_name))\n",
    "            os.system(f\"gunzip {os.path.join('genomes', file_name)}\")\n",
    "            \n",
    "            print(f\"Download complete on {user_input}\")\n",
    "            \n",
    "    print(f\"Successfully downloaded {num_genomes} genomes\")\n",
    "            \n",
    "query_database()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The k-mer distance estimation code\n",
    "\n",
    "The next cell contains the code for the k-mer distance estimation program. This includes code for extracting the k-mers for each of our downloaded genomes, computing distance based on the k-mers, and formatting the output into a heatmap. This cell prepares our program to run, but will not complete any calculations yet.\n",
    "\n",
    "Click on the cell below and hit the play button to run the cell of code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio.SeqIO, math, re, scipy, warnings\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.cluster.hierarchy as sch\n",
    "from statistics import pstdev\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# generates kmer dictionary from fasta formatted data\n",
    "def parse_fasta(fasta, k):\n",
    "    kmer_dict = {}\n",
    "\n",
    "    # parses fasta file using SeqRecord iterator\n",
    "    for record in Bio.SeqIO.parse(fasta, \"fasta\"):\n",
    "        seq = record.seq\n",
    "        regex = \"([A-Z][a-z]+)(( [a-z]+)+ ?[1-9]?)\"\n",
    "        match = re.search(regex, record.description)\n",
    "        try:\n",
    "            genus = match.group(1)\n",
    "            species = match.group(2)\n",
    "            organism = genus + species\n",
    "        except AttributeError:\n",
    "            organism = record.description\n",
    "\n",
    "        # Iterates over all combinations of k continuous nucleotides and adds the kmers to\n",
    "        # a dictionary with a counter of the number of instances of the kmers\n",
    "        for nuc_idx in range(len(seq) - (k - 1)):\n",
    "            kmer = seq[nuc_idx:(nuc_idx + k)]\n",
    "            kmer_dict.setdefault(kmer, 0)\n",
    "            kmer_dict[kmer] += 1\n",
    "            \n",
    "    print(\"Parsed {0} -------> Exracted Organism: {1}\".format(fasta, organism))\n",
    "    return (organism, kmer_dict)\n",
    "\n",
    "# compares the keys in two dictionaries, and adds the missing keys\n",
    "# from each into one another\n",
    "def compare_kmer(dict_1, dict_2):\n",
    "    set_1 = set(dict_1.keys())\n",
    "    set_2 = set(dict_2.keys())\n",
    "    diff_keys = list(set_1.symmetric_difference(set_2))\n",
    "    for key in diff_keys:\n",
    "        dict_1.setdefault(key, 0)\n",
    "        dict_2.setdefault(key, 0)\n",
    "\n",
    "# computes euclidean distance based on two kmer dictionaries\n",
    "def euclid_kmer(dict_1, dict_2):\n",
    "    euclid_sum = 0\n",
    "    for coord in dict_2:\n",
    "        euclid_sum += (dict_2[coord] - dict_1[coord])**2\n",
    "    return math.sqrt(euclid_sum)\n",
    "\n",
    "# normalizes distances in the output matrix to 1\n",
    "def normalize(distances):\n",
    "    all_values = []\n",
    "    for sub in distances:\n",
    "        for val in sub:\n",
    "            all_values.append(val)\n",
    "\n",
    "    minimum = min(all_values)\n",
    "    maximum = max(all_values)\n",
    "\n",
    "    # sets values in sub_lists to normalized values\n",
    "    for sub in distances:\n",
    "        for idx, val in enumerate(sub):\n",
    "            norm_val = (val - minimum)/(maximum - minimum)\n",
    "            sub[idx] = norm_val\n",
    "\n",
    "    return distances\n",
    "\n",
    "def matrix_output(samples, k):\n",
    "    ### {sample_N: {sample_N kmer dict}, ...} ###\n",
    "    distance_dict = {}\n",
    "    orgs = []\n",
    "\n",
    "    # pairs kmer dictionaries with organism names\n",
    "    for sample in samples:\n",
    "        org, distance_dict[sample] = parse_fasta(sample, k)\n",
    "        # adds organism to list for output header\n",
    "        if org not in orgs:\n",
    "                orgs.append(org)\n",
    "\n",
    "    ### {(sample_x, sample_y): dist, ...} ###\n",
    "    # stores previously computed distance values so program only\n",
    "    # has to run unique pairs of orgs\n",
    "    store_compare = {}\n",
    "\n",
    "    # distance matrix output; list of lists\n",
    "    distances = []\n",
    "    for sample_1 in samples:\n",
    "        sub_dist = []\n",
    "        for sample_2 in samples:\n",
    "            # if distance already computed, grab cached value\n",
    "            if (sample_2, sample_1) in store_compare:\n",
    "                sub_dist.append(store_compare[sample_2, sample_1])\n",
    "                continue\n",
    "            # if samples are the same, dist = 0\n",
    "            if sample_1 == sample_2:\n",
    "                sub_dist.append(0.00)\n",
    "                continue\n",
    "\n",
    "            # grabs kmer dictionaries from distance_dict\n",
    "            kmer_1 = distance_dict[sample_1]\n",
    "            kmer_2 = distance_dict[sample_2]\n",
    "\n",
    "            compare_kmer(kmer_1, kmer_2)\n",
    "            # calculates euclidean distance\n",
    "            dist = euclid_kmer(kmer_1, kmer_2)\n",
    "            # adds dist to cache dictionary\n",
    "            store_compare[(sample_1, sample_2)] = dist\n",
    "            sub_dist.append(dist)\n",
    "\n",
    "        distances.append(sub_dist)\n",
    "    return orgs, distances\n",
    "\n",
    "def make_table(k, orgs, distances):\n",
    "    # normalizes the data in distances for heatmap output\n",
    "    D = np.array(normalize(distances))\n",
    "    \n",
    "    fig = plt.figure(figsize=(8,8))\n",
    "    \n",
    "    # Dendrogram to right of heat map\n",
    "    ax1 = fig.add_axes([0.91, 0.1, 0.2, 0.6])\n",
    "    Y = sch.linkage(D, method='single')\n",
    "    Z1 = sch.dendrogram(Y, orientation='right')\n",
    "    ax1.set_xticks([]); ax1.set_yticks([])\n",
    "\n",
    "    # Dendrogram on top of heat map\n",
    "    ax2 = fig.add_axes([0.3, 0.71, 0.6, 0.2])\n",
    "    Y = sch.linkage(D, method='single')\n",
    "    Z2 = sch.dendrogram(Y)\n",
    "    ax2.set_xticks([]); ax2.set_yticks([])\n",
    "\n",
    "    # main heat-map\n",
    "    axmatrix = fig.add_axes([0.3, 0.1, 0.6, 0.6])\n",
    "    # orients species by distance from one another\n",
    "    idx1 = Z1['leaves']\n",
    "    idx2 = Z2['leaves']\n",
    "    D = D[idx1, :]\n",
    "    D = D[:, idx2]\n",
    "    # the actual heat-map\n",
    "    im = axmatrix.matshow(D, aspect='auto', origin='lower', cmap=\"RdBu\")\n",
    "    axmatrix.set_xticks([]); axmatrix.set_yticks([])\n",
    "\n",
    "    # labeling the x-axis\n",
    "    axmatrix.set_xticks(range(len(idx1)))\n",
    "    axmatrix.set_xticklabels([orgs[x] for x in idx1], minor=False)\n",
    "    axmatrix.xaxis.set_label_position('bottom')\n",
    "    axmatrix.xaxis.tick_bottom()\n",
    "\n",
    "    plt.xticks(rotation=-90, fontsize=8)\n",
    "\n",
    "    # labeling the y-axis\n",
    "    axmatrix.set_yticks(range(len(idx2)))\n",
    "    axmatrix.set_yticklabels([orgs[x] for x in idx2], minor=False)\n",
    "    axmatrix.yaxis.set_label_position('left')\n",
    "    axmatrix.yaxis.tick_left()\n",
    "\n",
    "    plt.show()\n",
    "    \n",
    "print(\"Successfully loaded k-mer distance estimation code\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running the k-mer distance estimation program\n",
    "\n",
    "The following code cell contains the code necessary to run the program itself. The program should update periodically at it goes through each genome and will output a heatmap at the end."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes = [genome.path for genome in os.scandir(\"genomes\")]\n",
    "organisms, distances = matrix_output(genomes, 4)\n",
    "make_table(4, organisms, distances)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyzing the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Back to main site](https://lawrencegordon.github.io/)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "e7370f93d1d0cde622a1f8e1c04877d8463912d04d973331ad4851f04de6915a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
